source/common/network/happy_eyeballs_connection_impl.h

```diff
@@ -0,0 +1,160 @@
+  void removeConnectionCallbacks(ConnectionCallbacks& cb) override;
+  void close(ConnectionCloseType type) override;
+  Event::Dispatcher& dispatcher() override;
+  uint64_t id() const override;
+  void hashKey(std::vector<uint8_t>& hash) const override;
+  void setConnectionStats(const ConnectionStats& stats) override;
+  void setDelayedCloseTimeout(std::chrono::milliseconds timeout) override;
+
+  // ScopeTrackedObject
+  void dumpState(std::ostream& os, int indent_level) const override;
+
+private:
+  // ConnectionCallbacks which will be set on an ClientConnection which
+  // sends connection events back to the HappyEyeballsConnectionImpl.
+  class ConnectionCallbacksWrapper : public ConnectionCallbacks {
+  public:
+    ConnectionCallbacksWrapper(HappyEyeballsConnectionImpl& parent, ClientConnection& connection)
+        : parent_(parent), connection_(connection) {}
+
+    void onEvent(ConnectionEvent event) override { parent_.onEvent(event, this); }
+
+    void onAboveWriteBufferHighWatermark() override {
+      parent_.onAboveWriteBufferHighWatermark(this);
+    }
+
+    void onBelowWriteBufferLowWatermark() override { parent_.onBelowWriteBufferLowWatermark(this); }
+
+    // Not needed? interesting.
+    ClientConnection& connection() { return connection_; }
+
+  private:
+    HappyEyeballsConnectionImpl& parent_;
+    ClientConnection& connection_;
+  };
+
+  std::unique_ptr<ClientConnection> createNextConnection();
+  void tryAnotherConnection();
+  void maybeScheduleNextAttempt();
+
+  void onEvent(ConnectionEvent event, ConnectionCallbacksWrapper* wrapper);
+
+  void onAboveWriteBufferHighWatermark(ConnectionCallbacksWrapper* wrapper);
+
+  void onBelowWriteBufferLowWatermark(ConnectionCallbacksWrapper* wrapper);
+
+  void onWriteBufferHighWatermark();
+
+  void cleanupWrapperAndConnection(ConnectionCallbacksWrapper* wrapper);
+
+  // State which needs to be applie to every connection attempt.
+  struct PerConnectionState {
+    absl::optional<bool> detect_early_close_when_read_disabled_;
+    absl::optional<bool> no_delay_;
+    absl::optional<const ConnectionStats*> connection_stats_;
+    absl::optional<uint32_t> buffer_limits_;
+  };
+
+  // State which needs to be saved and applied only to the final connection
+  // attempt.
+  struct PostConnectState {
+    std::vector<ConnectionCallbacks*> connection_callbacks_;
+    std::vector<ReadFilterSharedPtr> read_filters_;
+    absl::optional<Buffer::InstancePtr> write_buffer_;
+    absl::optional<bool> end_stream_;
+  };
+
+  Event::Dispatcher& dispatcher_;
+  const std::vector<Address::InstanceConstSharedPtr>& address_list_;
+  size_t next_address_ = 0;
+  Address::InstanceConstSharedPtr source_address_;
+  TransportSocketFactory& socket_factory_;
+  TransportSocketOptionsConstSharedPtr transport_socket_options_;
+  const ConnectionSocket::OptionsSharedPtr options_;
+
+  // Set of active connections.
+  std::vector<std::unique_ptr<ClientConnection>> connections_;
```

Comment:
You could use ClientConnectionPtr defined in https://github.com/envoyproxy/envoy/blob/0fae6970ddaf93f024908ba304bbd2b34e997a51/envoy/network/connection.h#L360 to get rid of a pair of brackets.